File: scripts/unique-dish-images.py
#!/usr/bin/env python3
"""Write a unique JPEG for every recipe that has no authentic photo."""
from __future__ import annotations

import hashlib
import json
import subprocess
import sys
from pathlib import Path

from PIL import Image, ImageEnhance, ImageOps

ROOT = Path(__file__).resolve().parents[1]
DISH = ROOT / "public" / "dishes"
UNIQUE = DISH / "u"
SKIP = {"hero"}


def authentic_ids() -> set[str]:
    return {p.stem for p in DISH.glob("*.jpg") if p.stem not in SKIP}


def hbytes(value: str) -> bytes:
    return hashlib.md5(value.encode("utf-8")).digest()


def variant(src: Path, slug: str, dest: Path) -> None:
    raw = hbytes(slug)
    im = Image.open(src).convert("RGB")
    width, height = im.size
    scale = 0.76 + (raw[0] / 255) * 0.18
    nw = max(8, int(width * scale))
    nh = max(8, int(height * scale))
    x = min(width - nw, int((width - nw) * raw[1] / 255))
    y = min(height - nh, int((height - nh) * raw[2] / 255))
    im = im.crop((x, y, x + nw, y + nh))
    if raw[3] > 200:
        im = ImageOps.mirror(im)
    angle = (raw[4] / 255) * 5.0 - 2.5
    im = im.rotate(angle, resample=Image.Resampling.BICUBIC, expand=False, fillcolor=(28, 22, 16))
    im = im.resize((540, 720), Image.Resampling.LANCZOS)
    im = ImageEnhance.Color(im).enhance(0.80 + raw[5] / 255 * 0.50)
    im = ImageEnhance.Contrast(im).enhance(0.86 + raw[6] / 255 * 0.38)
    im = ImageEnhance.Brightness(im).enhance(0.90 + raw[7] / 255 * 0.22)
    warm = Image.new(
        "RGB",
        im.size,
        (
            40 + raw[8] // 2,
            24 + raw[9] // 3,
            16 + raw[10] // 4,
        ),
    )
    im = Image.blend(im, warm, 0.035 + raw[11] / 255 * 0.07)
    dest.parent.mkdir(parents=True, exist_ok=True)
    im.save(dest, "JPEG", quality=70, optimize=True)


def dump_recipes() -> list[dict]:
    out = subprocess.check_output(
        ["npx", "tsx", "scripts/dump-recipe-images.ts"],
        cwd=ROOT,
    )
    return json.loads(out)


def main() -> int:
    auth = authentic_ids()
    recipes = dump_recipes()
    UNIQUE.mkdir(parents=True, exist_ok=True)
    written = 0
    skipped = 0
    missing_src = 0
    for recipe in recipes:
        slug = recipe["slug"]
        dest = UNIQUE / f"{slug}.jpg"
        if slug in auth:
            if dest.exists():
                dest.unlink()
            skipped += 1
            continue
        if dest.exists() and dest.stat().st_size > 8000:
            skipped += 1
            continue
        src_name = Path(str(recipe.get("source", ""))).name
        src = DISH / src_name
        if not src.exists():
            src = DISH / "hero.jpg"
            missing_src += 1
        variant(src, slug, dest)
        written += 1
    print(f"unique written={written} skipped={skipped} missing_src={missing_src} recipes={len(recipes)}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/unique-dish-images.mjs
#!/usr/bin/env node
import { createHash } from "node:crypto";
import { existsSync, mkdirSync, readFileSync, readdirSync, unlinkSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { Jimp } from "jimp";

const ROOT = join(dirname(fileURLToPath(import.meta.url)), "..");
const DISH = join(ROOT, "public/dishes");
const UNIQUE = join(DISH, "u");
const SOURCES = join(ROOT, "scripts/recipe-image-sources.json");

function authenticIds() {
  return new Set(
    readdirSync(DISH)
      .filter((name) => name.endsWith(".jpg") && name !== "hero.jpg")
      .map((name) => name.replace(/\.jpg$/, "")),
  );
}

function digest(slug) {
  return createHash("md5").update(slug).digest();
}

const sourceCache = new Map();

async function loadSource(path) {
  let img = sourceCache.get(path);
  if (!img) {
    img = await Jimp.read(path);
    sourceCache.set(path, img);
  }
  return img.clone();
}

async function variant(srcPath, slug, dest) {
  const raw = digest(slug);
  const im = await loadSource(srcPath);
  const width = im.width;
  const height = im.height;
  const scale = 0.76 + (raw[0] / 255) * 0.18;
  const nw = Math.max(8, Math.floor(width * scale));
  const nh = Math.max(8, Math.floor(height * scale));
  const x = Math.min(width - nw, Math.floor(((width - nw) * raw[1]) / 255));
  const y = Math.min(height - nh, Math.floor(((height - nh) * raw[2]) / 255));
  im.crop({ x, y, w: nw, h: nh });
  if (raw[3] > 200) im.flip({ horizontal: true });
  const angle = (raw[4] / 255) * 5 - 2.5;
  im.rotate(angle);
  im.resize({ w: 540, h: 720 });
  im.color([
    { apply: "saturate", params: [(raw[5] / 255) * 16] },
    { apply: "red", params: [(raw[8] / 255) * 10] },
    { apply: "blue", params: [(raw[9] / 255) * 8] },
  ]);
  im.contrast((raw[6] / 255) * 0.18 - 0.04);
  im.brightness((raw[7] / 255) * 0.12 - 0.04);
  await im.write(dest, { quality: 70 });
}

async function main() {
  const auth = authenticIds();
  const recipes = JSON.parse(readFileSync(SOURCES, "utf8"));
  mkdirSync(UNIQUE, { recursive: true });
  let written = 0;
  let skipped = 0;
  for (const recipe of recipes) {
    const slug = recipe.slug;
    const dest = join(UNIQUE, `${slug}.jpg`);
    if (auth.has(slug)) {
      if (existsSync(dest)) unlinkSync(dest);
      skipped += 1;
      continue;
    }
    if (existsSync(dest)) {
      skipped += 1;
      continue;
    }
    const srcName = String(recipe.source ?? "").replace(/^.*\//, "");
    let src = join(DISH, srcName);
    if (!existsSync(src)) src = join(DISH, "hero.jpg");
    await variant(src, slug, dest);
    written += 1;
    if (written % 200 === 0) console.log("unique", written);
  }
  console.log(`unique written=${written} skipped=${skipped} recipes=${recipes.length}`);
}

await main();
